File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章荷马史诗.py
#荷马史诗 lanqiaoOJ 1167
from heapq import *
n,k = map(int,input().split())
w = []
for i in range(n):   w.append(int(input()))
pq = [(i,0) for i in w]
while (n-1)%(k-1)!=0: pq += [(0,0)]; n += 1
heapify(pq)       #让列表具备堆特征
ans,x = 0,0
while len(pq)>1:
    s = 0
    for i in range(k):
        tmp = heappop(pq)
        s += tmp[0]
        x = max(x,tmp[1])
    ans += s
    heappush(pq,(s,x+1))
print(ans)
print(x+1)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章荷马史诗.java
//荷马史诗 lanqiaoOJ 1167


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;

public class Main{
    static int n,k;
    static long ans;
    public static void main(String[] args)throws IOException
    {
        BufferedReader sc = new BufferedReader(new InputStreamReader(System.in));
        PriorityQueue<Node> heap = new PriorityQueue<Node>((a,b)-> a.x==b.x?a.y-b.y:(a.x>b.x?1:-1));

        String[] line = sc.readLine().split(" ");
        n = Integer.parseInt(line[0]);
        k = Integer.parseInt(line[1]);
        int t = n;
        while (t-->0)
        {
            line = sc.readLine().split(" ");
            long a = Long.parseLong(line[0]);
            heap.offer(new Node(a,0));
        }
        int d = 0;
        if ((n-1)%(k-1)!=0)
            d = k-1-(n-1)%(k-1);
        while (d-->0)
        heap.offer(new Node(0,0));
        while (heap.size()>1)
        {
            long res1=0;
            int res2 = 0;
            for (int i=1;i<=k;i++)
            {
                long a = heap.peek().x;
                int b = heap.peek().y;
                heap.poll();
                res1+=a;
                res2 = Math.max(res2,b);
            }
            heap.offer(new Node(res1,res2+1));
            ans+=res1;
        }
        System.out.println(ans);
        System.out.println(heap.peek().y);
        
    }
}
class Node {
    long x;
    int y;
    public Node(long x,int y)
    {
        this.x = x;
        this.y = y;
    }

}